### PR TITLE
Identity Account

### PR DESCRIPTION
**Pull request by l1h3r**
Extends the `identity-account` module to include core functionality for creating and manipulating DID Documents.

Closes https://github.com/iotaledger/identity.rs/issues/139 and https://github.com/iotaledger/identity.rs/issues/140

### CODE REVIEW

**Review by JelleMillenaar — CHANGES_REQUESTED** [review]

File: identity-account/examples/signing.rs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    .fragment("key-1")
+    .finish()?;
+
+  // Process the command and update the identity state.
+  account.update(snapshot.id(), command).await?;
+
+  let mut credential: Credential = Credential::builder(Default::default())
+    .issuer(Url::parse("https://example.com")?)
+    .type_("ExampleCredential")
+    .subject(json!({"foo": "bar"}).serde_into()?)
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply identity-account/examples/signing.rs]: Replaced with `UniversityDegreeCredential` 😆

File: identity-account/examples/methods.rs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  // Fetch and log the DID Document from the Tangle
+  println!(
+    "[Example] Tangle Document (1) = {:#?}",
+    account.resolve(snapshot.id()).await?
+  );
+
+  // Add another Ed25519 verification method to the identity
+  let command: Command = Command::create_method()
+    .type_(MethodType::Ed25519VerificationKey2018)
+    .scope(MethodScope::VerificationMethod)
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply identity-account/examples/methods.rs]: done for `MethodType` and was apparently already done for `MethodScope`. left out `fragment` as we discussed earlier

File: identity-account/examples/config.rs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+use identity_account::account::Config;
+use identity_account::error::Result;
+use identity_account::storage::MemStore;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+  pretty_env_logger::init();
+
+  let config: Config = Config::new()
+    .autosave(AutoSave::Never) // never auto-save. rely on the drop save
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply identity-account/examples/config.rs:17]: just locally for now - publishing to the tangle isn't configurable at the moment but will do in a follow-up PR

File: identity-account/examples/signing.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    .build()?;
+
+  // Sign the credential with the previously created verification method
+  account.sign(snapshot.id(), "key-1", &mut credential).await?;
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply identity-account/examples/signing.rs]: Updated the examples the use the DID instead as discussed earlier and may expand with a follow-up PR to add a context-based API

File: identity-account/examples/basic.rs [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+#[tokio::main]
+async fn main() -> Result<()> {
+  pretty_env_logger::init();
+
+  // Create an in-memory storage instance for the account
+  let storage: MemStore = MemStore::new();
+
+  // Create a new Account with the default configuration
+  let account: Account<MemStore> = Account::new(storage).await?;
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply identity-account/examples/basic.rs:21]: `network` is currently configurable through `IdentityCreate` - we have a map of `{network => client}` and currently just create `Client`s on-demand with default settings.

Once `identity::iota::Client` is more configurable we can start adding those options to `identity::account::Config`

File: identity-account/examples/stronghold.rs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  println!("[Example] Local Document = {:#?}", snapshot.identity().to_document()?);
+  println!("[Example] Local Document List = {:#?}", account.list().await);
+
+  // Fetch the DID Document from the Tangle
+  let resolved: Document = account.resolve(snapshot.id()).await?;
+
+  println!("[Example] Tangle Document = {:#?}", resolved);
+
+  // Create another new Identity
+  let snapshot: IdentitySnapshot = account.create(Default::default()).await?;
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply identity-account/examples/stronghold.rs]: Updated so this is more explicit

File: identity-account/examples/methods.rs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+  // Create a new Identity with default settings
+  let snapshot: IdentitySnapshot = account.create(Default::default()).await?;
+
+  // Retrieve the DID from the newly created Identity state.
+  let document: &IotaDID = snapshot.identity().try_document()?;
+
+  // Add a new Ed25519 (defualt) verification method to the identity - the
+  // verification method is included as an embedded authentication method.
+  let command: Command = Command::create_method()
```
Comment (JelleMillenaar) [inline identity-account/examples/methods.rs:32]: Would it perhaps be possible to keep track of the command inside the account as well? Then we can have an array of unpublished commands and when you call `account.update(DID).await?;` it publishes all the commands in a single message. This is more efficient on the Tangle as you can easily group commands and easier for the developers as they don't have to keep track of commands.

File: identity-account/examples/methods.rs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  let storage: MemStore = MemStore::new();
+
+  // Create a new Account with the default configuration
+  let account: Account<MemStore> = Account::new(storage).await?;
+
+  // Create a new Identity with default settings
+  let snapshot: IdentitySnapshot = account.create(Default::default()).await?;
+
+  // Retrieve the DID from the newly created Identity state.
+  let document: &IotaDID = snapshot.identity().try_document()?;
```
Comment (JelleMillenaar) [inline identity-account/examples/methods.rs:28]: DID or DID Document? The naming suggests DID Document, but the type suggests a DID ;D

File: identity-account/examples/basic.rs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+use identity_account::identity::IdentitySnapshot;
+use identity_account::storage::MemStore;
+use identity_iota::did::Document;
+
+#[tokio::main]
+async fn main() -> Result<()> {
+  pretty_env_logger::init();
+
+  // Create an in-memory storage instance for the account
+  let storage: MemStore = MemStore::new();
```
[the remark this answers is not in the document]
Comment (l1h3r) [reply identity-account/examples/basic.rs]: new pattern is much nicer

https://github.com/iotaledger/identity.rs/pull/151/files#diff-5bb63b6b225d925f3c7dffc2f304a44ecab0a58fef284c80b81f5231243d0d15R16

File: identity-account/examples/methods.rs (hunk 7 above)
Comment (l1h3r) [reply identity-account/examples/methods.rs]: follow-up PR 😛

**Review by JelleMillenaar — APPROVED** [review]
LGTY

**Pull request merged**